persist/sqlite/consensus_refactored_test.go

```diff
@@ -1703,3 +1713,248 @@ func TestFileContractMultipleRevisions(t *testing.T) {
+	n.mineTransactions(t, txn2)
+
+	fceRevision1 := fce
+	fceRevision1.RevisionNumber = fcRevision1.RevisionNumber
+	fceRevision1.TransactionID = txn2.ID()
+
+	// either key should be associated with the contract
+	n.assertContractsKey(t, pk1.PublicKey(), fceRevision1)
+	n.assertContractsKey(t, pk2.PublicKey(), fceRevision1)
+
+	n.revertBlock(t)
+
+	// if we revert we should keep the keys.  only reason to change them is if
+	// we change the UnlockHash and can get the keys from the UnlockConditions
+	// in a future revision
+	n.assertContractsKey(t, pk1.PublicKey(), fce)
+	n.assertContractsKey(t, pk2.PublicKey(), fce)
+
+	n.revertBlock(t)
+
+	n.assertContractsKey(t, pk1.PublicKey())
+	n.assertContractsKey(t, pk2.PublicKey())
+}
+
+func TestMetrics(t *testing.T) {
+	pk1 := types.GeneratePrivateKey()
+	uc1 := types.StandardUnlockConditions(pk1.PublicKey())
+	addr1 := uc1.UnlockHash()
+
+	n := newTestChain(t, false, func(network *consensus.Network, genesisBlock types.Block) {
+		genesisBlock.Transactions[0].SiacoinOutputs[0].Address = addr1
+	})
+	val := n.genesis().Transactions[0].SiacoinOutputs[0].Value
+
+	assertMetrics := func(expected explorer.Metrics) {
+		t.Helper()
+
+		got, err := n.db.Metrics(n.tipState().Index.ID)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		cs := n.tipState()
+		testutil.Equal(t, "Index", cs.Index, got.Index)
+		testutil.Equal(t, "Difficulty", cs.Difficulty, got.Difficulty)
+		testutil.Equal(t, "SiafundTaxRevenue", cs.SiafundTaxRevenue, got.SiafundTaxRevenue)
+		testutil.Equal(t, "NumLeaves", cs.Elements.NumLeaves, got.NumLeaves)
+		testutil.Equal(t, "ActiveContracts", expected.ActiveContracts, got.ActiveContracts)
+		testutil.Equal(t, "FailedContracts", expected.FailedContracts, got.FailedContracts)
+		testutil.Equal(t, "SuccessfulContracts", expected.SuccessfulContracts, got.SuccessfulContracts)
+		testutil.Equal(t, "StorageUtilization", expected.StorageUtilization, got.StorageUtilization)
+		testutil.Equal(t, "CirculatingSupply", expected.CirculatingSupply, got.CirculatingSupply)
+		testutil.Equal(t, "ContractRevenue", expected.ContractRevenue, got.ContractRevenue)
+	}
+
+	var circulatingSupply types.Currency
+	for _, txn := range n.genesis().Transactions {
+		for _, sco := range txn.SiacoinOutputs {
+			circulatingSupply = circulatingSupply.Add(sco.Value)
+		}
+	}
+	metricsGenesis := explorer.Metrics{
+		CirculatingSupply: circulatingSupply,
+	}
+	assertMetrics(metricsGenesis)
+
+	// form two contracts
+	fc := prepareContract(addr1, n.tipState().Index.Height+3)
+	txn1 := types.Transaction{
+		SiacoinInputs: []types.SiacoinInput{{
+			ParentID:         n.genesis().Transactions[0].SiacoinOutputID(0),
+			UnlockConditions: uc1,
+		}},
+		SiacoinOutputs: []types.SiacoinOutput{{
+			Address: addr1,
+			Value:   val.Sub(fc.Payout.Mul64(2)),
+		}},
+		FileContracts: []types.FileContract{fc, fc},
+	}
+	testutil.SignTransaction(n.tipState(), pk1, &txn1)
+
+	if subsidy, ok := n.tipState().FoundationSubsidy(); ok {
```

Comment:
`nit` I'd mine an empty block after asserting the circulating supply, that way testing that funds locked away in contract impact the circulating supply is slightly cleaner